lib/base-service.ts

```diff
@@ -0,0 +1,185 @@
+   * @param {UserOptions} options
+   * @param {OutgoingHttpHeaders} [options.headers]
+   * @param {string} [options.url] - override default service base url
+   * @private
+   * @abstract
+   * @constructor
+   * @throws {Error}
+   * @returns {BaseService}
+   */
+  constructor(userOptions: UserOptions) {
+    if (!(this instanceof BaseService)) {
+      throw new Error(
+        'the "new" keyword is required to create service instances'
+      );
+    }
+
+    const _options = {} as BaseServiceOptions;
+    const options = extend({}, userOptions);
+
+    // for compatibility
+    if (options.url && !options.serviceUrl) {
+      options.serviceUrl = options.url;
+    }
+
+    if (options.serviceUrl) {
+      _options.serviceUrl = stripTrailingSlash(options.serviceUrl);
+    }
+
+    // check serviceUrl for common user errors
+    const credentialProblems = checkCredentials(options, ['serviceUrl']);
+    if (credentialProblems) {
+      throw credentialProblems;
+    }
+
+    // if disableSslVerification is not explicity set to the boolean value `true`,
+    // force it to be false
+    if (options.disableSslVerification !== true) {
+      options.disableSslVerification = false;
+    }
+
+    const serviceClass = this.constructor as typeof BaseService;
+    this.baseOptions = extend(
+      { qs: {}, serviceUrl: serviceClass.URL },
+      options,
+      this.readOptionsFromExternalConfig(),
+      _options
+    );
+
+    this.requestWrapperInstance = new RequestWrapper(this.baseOptions);
+
+    // enforce that an authenticator is set
+    if (!options.authenticator) {
+      throw new Error('Authenticator must be set.');
+    }
+
+    this.authenticator = options.authenticator;
+  }
+
+  /**
+   * Get the instance of the authenticator set on the service.
+   *
+   * @returns {Authenticator}
+   */
+  public getAuthenticator(): any {
+    return this.authenticator;
+  }
+
+  /**
+   * Set the service URL to send requests to.
+   *
+   * @param {string} the base URL for the service
+   */
+  public setServiceUrl(url: string): void {
+    this.baseOptions.serviceUrl = url;
+  }
+
+  /**
+   * Wrapper around `sendRequest` that enforces the request will be authenticated.
+   *
+   * @param {Object} parameters - service request options passed in by user.
+   * @param {string} parameters.options.method - the http method.
+   * @param {string} parameters.options.url - the path portion of the URL to be appended to the serviceUrl
+   * @param {string} parameters.options.path - the path to be appended to the service URL.
+   * @param {string} parameters.options.qs - the querystring to be included in the URL.
+   * @param {string} parameters.options.body - the data to be sent as the request body.
+   * @param {Object} parameters.options.form - an object containing the key/value pairs for a www-form-urlencoded request.
+   * @param {Object} parameters.options.formData - an object containing the contents for a multipart/form-data request.
```

Comment:
The convention is that we designate optional parameters by putting them in brackets (`[]`), right?  If so, should we indicate this for `parameters.options.qs`, `parameters.options.body`, `parameters.options.form`, and `parameters.options.formData`?

Reply:
Yes, that's a good call. I think `parameters.options.path` also. Those are path parameters - I will change the description as well.

Reply:
fixed